Add PBKDF1-SHA1 / m32900

PenguinKeeper7 opened the pull request:
Satisfy https://github.com/hashcat/hashcat/issues/4089
```
$ ./tools/test.sh -m 32900 -a all -t all -o win -P
[ test_1729202194 ] > Init test for hash type 32900.
[ test_1729202194 ] [ Type 32900, Attack 0, Mode single, Device-Type Gpu, Kernel-Type Pure, Vector-Width 1 ] > OK : 0/8 not found, 0/8 not matched, 0/8 timeout, 0/8 skipped
[ test_1729202194 ] [ Type 32900, Attack 0, Mode multi,  Device-Type Gpu, Kernel-Type Pure, Vector-Width 1 ] > OK : 0/1 not found, 0/1 not matched, 0/1 timeout, 0/1 skipped
[ test_1729202194 ] [ Type 32900, Attack 1, Mode single, Device-Type Gpu, Kernel-Type Pure, Vector-Width 1 ] > OK : 0/7 not found, 0/7 not matched, 0/7 timeout, 0/7 skipped
[ test_1729202194 ] [ Type 32900, Attack 1, Mode multi,  Device-Type Gpu, Kernel-Type Pure, Vector-Width 1 ] > OK : 0/1 not found, 0/1 not matched, 0/1 timeout, 0/1 skipped
[ test_1729202194 ] [ Type 32900, Attack 3, Mode single, Device-Type Gpu, Kernel-Type Pure, Vector-Width 1 ] > OK : 0/7 not found, 0/7 not matched, 0/7 timeout, 0/7 skipped
[ test_1729202194 ] [ Type 32900, Attack 3, Mode multi,  Device-Type Gpu, Kernel-Type Pure, Vector-Width 1 ] > OK : 0/1 not found, 0/1 not matched, 0/1 timeout, 0/1 skipped
[ test_1729202194 ] [ Type 32900, Attack 6, Mode single, Device-Type Gpu, Kernel-Type Pure, Vector-Width 1 ] > Warning : 0/7 not found, 0/7 not matched, 3/7 timeout, 0/7 skipped
[ test_1729202194 ] [ Type 32900, Attack 6, Mode multi,  Device-Type Gpu, Kernel-Type Pure, Vector-Width 1 ] > OK : 0/7 not found, 0/7 not matched, 0/7 timeout, 0/7 skipped
[ test_1729202194 ] [ Type 32900, Attack 7, Mode single, Device-Type Gpu, Kernel-Type Pure, Vector-Width 1 ] > OK : 0/7 not found, 0/7 not matched, 0/7 timeout, 0/7 skipped
[ test_1729202194 ] [ Type 32900, Attack 7, Mode multi,  Device-Type Gpu, Kernel-Type Pure, Vector-Width 1 ] > OK : 0/7 not found, 0/7 not matched, 0/7 timeout, 0/7 skipped
[ test_1729202194 ] [ Type 32900, Attack 0, Mode single, Device-Type Gpu, Kernel-Type Pure, Vector-Width 4 ] > OK : 0/8 not found, 0/8 not matched, 0/8 timeout, 0/8 skipped
[ test_1729202194 ] [ Type 32900, Attack 0, Mode multi,  Device-Type Gpu, Kernel-Type Pure, Vector-Width 4 ] > OK : 0/1 not found, 0/1 not matched, 0/1 timeout, 0/1 skipped
[ test_1729202194 ] [ Type 32900, Attack 1, Mode single, Device-Type Gpu, Kernel-Type Pure, Vector-Width 4 ] > OK : 0/7 not found, 0/7 not matched, 0/7 timeout, 0/7 skipped
[ test_1729202194 ] [ Type 32900, Attack 1, Mode multi,  Device-Type Gpu, Kernel-Type Pure, Vector-Width 4 ] > OK : 0/1 not found, 0/1 not matched, 0/1 timeout, 0/1 skipped
[ test_1729202194 ] [ Type 32900, Attack 3, Mode single, Device-Type Gpu, Kernel-Type Pure, Vector-Width 4 ] > OK : 0/7 not found, 0/7 not matched, 0/7 timeout, 0/7 skipped
[ test_1729202194 ] [ Type 32900, Attack 3, Mode multi,  Device-Type Gpu, Kernel-Type Pure, Vector-Width 4 ] > OK : 0/1 not found, 0/1 not matched, 0/1 timeout, 0/1 skipped
[ test_1729202194 ] [ Type 32900, Attack 6, Mode single, Device-Type Gpu, Kernel-Type Pure, Vector-Width 4 ] > Warning : 0/7 not found, 0/7 not matched, 3/7 timeout, 0/7 skipped
[ test_1729202194 ] [ Type 32900, Attack 6, Mode multi,  Device-Type Gpu, Kernel-Type Pure, Vector-Width 4 ] > OK : 0/7 not found, 0/7 not matched, 0/7 timeout, 0/7 skipped
[ test_1729202194 ] [ Type 32900, Attack 7, Mode single, Device-Type Gpu, Kernel-Type Pure, Vector-Width 4 ] > OK : 0/7 not found, 0/7 not matched, 0/7 timeout, 0/7 skipped
[ test_1729202194 ] [ Type 32900, Attack 7, Mode multi,  Device-Type Gpu, Kernel-Type Pure, Vector-Width 4 ] > OK : 0/7 not found, 0/7 not matched, 0/7 timeout, 0/7 skipped
```

PenguinKeeper7 committed:
Add PBKDF1-SHA1 / m32900

Satisfy https://github.com/hashcat/hashcat/issues/4089

matrix commented:
Hi @PenguinKeeper7, I think this module must be ATTACK_EXEC_OUTSIDE_KERNEL and you need change the kernel implementation. Is there a particular reason why you wanted to implement it differently?
Thanks

PenguinKeeper7 commented:
On a 4090, I get 42MH/s (Obviously iterations can differ etc) which is rather fast so I just felt like it was more appropriate for exec inside but it is rather borderline and I really don't think exec outside would make any kind of positive impact.

matrix commented:
other than that, try testing it using the CPU and you will find a problem to solve.

PenguinKeeper7 committed:
Fix vectorisation issues

PenguinKeeper7 committed:
Old code cleanup

PenguinKeeper7 commented:
```
$ ./tools/test.sh -m 32900 -a all -t all -o win -P -D 1,2
[ test_1745673068 ] > Init test for hash type 32900.
[ test_1745673068 ] [ Type 32900, Attack 0, Mode single, Device-Type Cpu + Gpu, Kernel-Type Pure, Vector-Width 1 ] > OK : 0/8 not found, 0/8 not matched, 0/8 timeout, 0/8 skipped
[ test_1745673068 ] [ Type 32900, Attack 0, Mode multi,  Device-Type Cpu + Gpu, Kernel-Type Pure, Vector-Width 1 ] > OK : 0/1 not found, 0/1 not matched, 0/1 timeout, 0/1 skipped
[ test_1745673068 ] [ Type 32900, Attack 1, Mode single, Device-Type Cpu + Gpu, Kernel-Type Pure, Vector-Width 1 ] > OK : 0/7 not found, 0/7 not matched, 0/7 timeout, 0/7 skipped
[ test_1745673068 ] [ Type 32900, Attack 1, Mode multi,  Device-Type Cpu + Gpu, Kernel-Type Pure, Vector-Width 1 ] > OK : 0/1 not found, 0/1 not matched, 0/1 timeout, 0/1 skipped
[ test_1745673068 ] [ Type 32900, Attack 3, Mode single, Device-Type Cpu + Gpu, Kernel-Type Pure, Vector-Width 1 ] > OK : 0/7 not found, 0/7 not matched, 0/7 timeout, 0/7 skipped
[ test_1745673068 ] [ Type 32900, Attack 3, Mode multi,  Device-Type Cpu + Gpu, Kernel-Type Pure, Vector-Width 1 ] > OK : 0/1 not found, 0/1 not matched, 0/1 timeout, 0/1 skipped
[ test_1745673068 ] [ Type 32900, Attack 6, Mode single, Device-Type Cpu + Gpu, Kernel-Type Pure, Vector-Width 1 ] > Warning : 0/7 not found, 0/7 not matched, 2/7 timeout, 0/7 skipped
[ test_1745673068 ] [ Type 32900, Attack 6, Mode multi,  Device-Type Cpu + Gpu, Kernel-Type Pure, Vector-Width 1 ] > OK : 0/7 not found, 0/7 not matched, 0/7 timeout, 0/7 skipped
[ test_1745673068 ] [ Type 32900, Attack 7, Mode single, Device-Type Cpu + Gpu, Kernel-Type Pure, Vector-Width 1 ] > OK : 0/7 not found, 0/7 not matched, 0/7 timeout, 0/7 skipped
[ test_1745673068 ] [ Type 32900, Attack 7, Mode multi,  Device-Type Cpu + Gpu, Kernel-Type Pure, Vector-Width 1 ] > OK : 0/7 not found, 0/7 not matched, 0/7 timeout, 0/7 skipped
[ test_1745673068 ] [ Type 32900, Attack 0, Mode single, Device-Type Cpu + Gpu, Kernel-Type Pure, Vector-Width 4 ] > OK : 0/8 not found, 0/8 not matched, 0/8 timeout, 0/8 skipped
[ test_1745673068 ] [ Type 32900, Attack 0, Mode multi,  Device-Type Cpu + Gpu, Kernel-Type Pure, Vector-Width 4 ] > OK : 0/1 not found, 0/1 not matched, 0/1 timeout, 0/1 skipped
[ test_1745673068 ] [ Type 32900, Attack 1, Mode single, Device-Type Cpu + Gpu, Kernel-Type Pure, Vector-Width 4 ] > OK : 0/7 not found, 0/7 not matched, 0/7 timeout, 0/7 skipped
[ test_1745673068 ] [ Type 32900, Attack 1, Mode multi,  Device-Type Cpu + Gpu, Kernel-Type Pure, Vector-Width 4 ] > OK : 0/1 not found, 0/1 not matched, 0/1 timeout, 0/1 skipped
[ test_1745673068 ] [ Type 32900, Attack 3, Mode single, Device-Type Cpu + Gpu, Kernel-Type Pure, Vector-Width 4 ] > OK : 0/7 not found, 0/7 not matched, 0/7 timeout, 0/7 skipped
[ test_1745673068 ] [ Type 32900, Attack 3, Mode multi,  Device-Type Cpu + Gpu, Kernel-Type Pure, Vector-Width 4 ] > OK : 0/1 not found, 0/1 not matched, 0/1 timeout, 0/1 skipped
[ test_1745673068 ] [ Type 32900, Attack 6, Mode single, Device-Type Cpu + Gpu, Kernel-Type Pure, Vector-Width 4 ] > Warning : 0/7 not found, 0/7 not matched, 2/7 timeout, 0/7 skipped
[ test_1745673068 ] [ Type 32900, Attack 6, Mode multi,  Device-Type Cpu + Gpu, Kernel-Type Pure, Vector-Width 4 ] > OK : 0/7 not found, 0/7 not matched, 0/7 timeout, 0/7 skipped
[ test_1745673068 ] [ Type 32900, Attack 7, Mode single, Device-Type Cpu + Gpu, Kernel-Type Pure, Vector-Width 4 ] > OK : 0/7 not found, 0/7 not matched, 0/7 timeout, 0/7 skipped
[ test_1745673068 ] [ Type 32900, Attack 7, Mode multi,  Device-Type Cpu + Gpu, Kernel-Type Pure, Vector-Width 4 ] > OK : 0/7 not found, 0/7 not matched, 0/7 timeout, 0/7 skipped
```

matrix commented:
ok, the last thing it's porting to ATTACK_EXEC_OUTSIDE_KERNEL

PenguinKeeper7 commented:
I don't see a reason to. If anything, it'd just slow it down and even by the plugin development guide's general guidelines:
```
Expected less than 10 million guesses per second per GPU? -> slow kernel
```
Obviously it's just more of a vague suggestion than a hard limit but we're 4x over that recommendation so...¯\_(ツ)_/¯

matrix commented:
> Obviously it's just more of a vague suggestion than a hard limit but we're 4x over that recommendation so

OK, let's say for now the important thing is that it works. Then will be evaluated by atom.
Thanks!

PenguinKeeper7 committed:
Update module for the Assimilation Bridge

https://github.com/hashcat/hashcat/commit/ceb5ff5641c0c7c6452c34e9f942acd0465e4849#diff-415ad91017d1247a6dd408ec0006bd88a94425579228a7b5d394f2d86f0bd7da

PenguinKeeper7 commented:
Updated for the Assimilation Bridge module changes https://github.com/hashcat/hashcat/pull/4092/commits/ab3b971144d47eb63a6fbabfd1c9f63fa9683268

jsteube commented:
This is a cool edge case, because of the 1000 iterations. Reminds me on md5crypt -m500, which also has 1000 iterations, and it's implemented as ATTACK_EXEC_OUTSIDE_KERNEL.

I like the out of the box thinking, but have a closer look at these numbers from a simple -b -m 32900 run:

```
Speed.#01........: 41391.4 kH/s (75.29ms) @ Accel:3 Loops:8 Thr:1024 Vec:1
```

In this case hashcat had almost zero chance to autotune accel and loops because of the high runtime of the kernel. In the past, that was an absolute no go since all driver watchdogs would instantly restart the driver (bad!). These numbers are from a 4090, and the slower the GPU is, the worse this is. 

I think it makes sense to implement this as ATTACK_EXEC_OUTSIDE_KERNEL, but we might to  change our mind if the speed loss is too intense. Keep in mind, you also get something for it, even if it's slower. The amplifier requirement is gone, since 1000 iterations should be almost enough to make pcie bottleneck almost irrelevant.

Making a theoretical decision is hard. I think we need to test, and then decide with numbers, maybe on a more likely gpu than a 4090.

Also please add the new KERNEL_FA macro to the kernel declaration.

PenguinKeeper7 committed:
Merge branch 'hashcat:master' into pbkdf1-sha1

PenguinKeeper7 commented:
Machine 1 (3070, PCIE 8x4.0)
Exec inside - 10231.9 kH/s (73.28ms) @ Accel:4 Loops:8 Thr:512 Vec:1
Exec outside - 9887.8 kH/s (97.88ms) @ Accel:43 Loops:500 Thr:1024 Vec:1
3.48% faster with exec inside

Machine 2 (4090, PCIE 16x4.0)
Exec inside - 43675.2 kH/s (71.49ms) @ Accel:3 Loops:16 Thr:512 Vec:1
Exec outside - 41546.6 kH/s (95.13ms) @ Accel:128 Loops:500 Thr:512 Vec:1
5.12% faster with exec inside

Certainly better hashrate than I thought there'd be. My machines, of course have quite decent PCIE connections, so this may be exaggerated more than a user with worse connections. Also, there doesn't really seem to be much, if any autotuning issues around lower vs higher-end hardware, especially for the -n value but, as Atom said, this should probably be verified on lower-end machines, like mining rigs with terrible PCIE or lower-end cards that may struggle harder than my 3070 to autotune

PenguinKeeper7 commented:
I asked around and the weakest hardware I got was "Intel HD Graphics 530" (not controlled by me, unknown connection speed)
Machine 3
Exec inside - 120.6 kH/s (73.03ms) @ Accel:3 Loops:1 Thr:128 Vec:1
Exec outside - 129.6 kH/s (89.91ms) @ Accel:8 Loops:500 Thr:128 Vec:1
This time, 7.46% faster with outside, which is new but not a huge difference in speed. So even with hardware that is over 300x weaker than a 4090, the exec inside option still holds up fairly well so I imagine the actual crossover point in speed is around NVIDIA 500/700 series but it may be best to just go with the exec outside option for stability, even if it's slower on what I imagine are 99% of user rigs.

Thoughts? Happy to go with either option as both options are fully developed at this point, anyway.

jsteube commented:
OK, that's good news and pretty much what we expected. Could you please update the PR to include only the OUTSIDE kernel? It's definitely useful to have the INSIDE kernel as well, but we shouldn't include it in the default distribution to avoid confusing users. Feel free to share it on the forum as a showcase for developer so it's not lost!

PenguinKeeper7 committed:
Convert to ATTACK_EXEC_OUTSIDE_KERNEL

PenguinKeeper7 commented:
Replaced with the exec outside plugin. Tests continue to pass
```
$ tools/test.sh -m 32900 -a all -t all -o win -P
[ test_1752692256 ] > Init test for hash type 32900.
[ test_1752692256 ] [ Type 32900, Attack 0, Mode single, Device-Type Gpu, Kernel-Type Pure, Vector-Width 1 ] > OK : 0/8 not found, 0/8 not matched, 0/8 timeout, 0/8 skipped
[ test_1752692256 ] [ Type 32900, Attack 0, Mode multi,  Device-Type Gpu, Kernel-Type Pure, Vector-Width 1 ] > OK : 0/1 not found, 0/1 not matched, 0/1 timeout, 0/1 skipped
[ test_1752692256 ] [ Type 32900, Attack 0, Mode single, Device-Type Gpu, Kernel-Type Pure, Vector-Width 4 ] > OK : 0/8 not found, 0/8 not matched, 0/8 timeout, 0/8 skipped
[ test_1752692256 ] [ Type 32900, Attack 0, Mode multi,  Device-Type Gpu, Kernel-Type Pure, Vector-Width 4 ] > OK : 0/1 not found, 0/1 not matched, 0/1 timeout, 0/1 skipped
```

Forum post - https://hashcat.net/forum/thread-13315.html

jsteube commented:
Cool, PBKDF1, thanks!
